### PR TITLE
Enable offline cache

### LINKED ISSUE #3812: Login while offline

### PR DESCRIPTION
**Pull request by charlag**
Close #3812

### CODE REVIEW

**Comment by ganthern** [comment]
* [ ] smoke tested
* [ ] good commit message
* [ ] doc comments where necessary
* [ ] commits are squashed appropriately
* [ ] sufficient test notes
* [ ] changes reviewed
* [ ] translation keys are in Phrase

**Comment by ganthern** [comment]
I'm getting a consistent test failure:
```
OfflineDb  > Test encryption > can't read from existing database if password is wrong:
AssertThrows failed: Expected a function Error() { [native code] } to be thrown, but nothing was

```

**Comment by ganthern** [comment]
Getting errors when logged into two different accounts and then deleting the credentials to one of them from a third window:

The window that was logged into the deleted account:
```
Feedback v3.94.6 - Error - UNKNOWN - linux



 Client: linux
 Type: UNKNOWN
 Tutanota version: 3.94.6
 Timestamp (UTC): Thu, 07 Apr 2022 10:24:36 GMT
 User agent:
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) tutanota-desktop-debug/3.94.6 Chrome/98.0.4758.109 Electron/17.2.0 Safari/537.36
Error
 Error message: Db for user MfCQs9T----0 is not open. must call openDataBaseForUser first :)
Stacktrace: 
Error: Db for user MfCQs9T----0 is not open. must call openDataBaseForUser first :)
    at OfflineDbFacade.getDbForUserId (nollup-int:///src/desktop/db/OfflineDbFacade.ts:71:19)
    at OfflineDbFacade.putLastBatchIdForGroup (nollup-int:///src/desktop/db/OfflineDbFacade.ts:60:21)
    at eval (nollup-int:///src/api/common/WorkerProxy.ts:34:34)
    at IPC._invokeMethod (nollup-int:///src/desktop/IPC.ts:300:58)
    at eval (nollup-int:///src/desktop/IPC.ts:56:22)
    at node:electron/js2c/browser_init:193:579
    at EventEmitter.<anonymous> (node:electron/js2c/browser_init:161:10433)
    at EventEmitter.emit (node:events:390:28)
```

The window that was logged into the other account:
```
Feedback v3.94.6 - Error - PREMIUM - linux
Client: linux
Type: PREMIUM
Tutanota version: 3.94.6
Timestamp (UTC): Thu, 07 Apr 2022 10:24:43 GMT
User agent:
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) tutanota-desktop-debug/3.94.6 Chrome/98.0.4758.109 Electron/17.2.0 Safari/537.36
Error
Error message: ERR_FAILED (-2) loading 'file:///home/nig/dev/repositories/tutanota-3/build/index-desktop.html?noAutoLogin=true&platformId=linux&theme=%7B%22themeId%22%3A%22light%22%2C%22logo%22%3A%22%3Csvg+xmlns%3D%5C%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%5C%22+xml%3Aspace%3D%5C%22preserve%5C%22+preserveAspectRatio%3D%5C%22xMinYMid%5C%22+viewBox%3D%5C%220+0+1054.473+236.49%5C%22%3E%3Cg+transform%3D%5C%22translate%28-132.628+-222.799%29%5C%22+style%3D%5C%22fill%3A%234a4a4a%3Bfill-opacity%3A1%5C%22%3E%5Cn%5Ct%5Ct%5Ct%3Cpath+d%3D%5C%22M494.016+299.858H451.32v-8.082h94.178v8.082h-42.696v121.939h-8.785V299.858zM543.209+396.496v-66.943h8.434v65.889c0+14.057+6.15+20.908+20.206+20.908+12.827+0+23.544-6.676+34.965-17.57v-69.228h8.434v92.245h-8.434v-15.285c-10.191+9.662-22.139+17.219-35.668+17.219-19.152-.001-27.937-10.543-27.937-27.235zM652.319+401.24v-64.133h-16.165v-7.555h16.165v-33.56h8.434v33.56h24.071v7.555h-24.071v62.902c0+9.84+3.339+15.463+14.935+15.463+3.515+0+7.028-.352+9.664-1.23v7.906c-2.987.527-6.501.879-10.366.879-15.287+0-22.667-6.15-22.667-21.787zM703.27+400.537c0-17.57+14.935-30.749+63.43-38.831v-5.623c0-14.056-7.38-21.084-20.03-21.084-15.287+0-25.478+5.974-35.845+15.287l-4.919-5.271c11.421-10.542+23.192-17.395+40.938-17.395+19.152+0+28.289+10.894+28.289+27.937v43.399c0+11.246.703+18.097+2.636+22.841h-8.961c-1.23-3.865-2.108-8.434-2.108-13.705-11.245+9.664-23.545+15.287-37.426+15.287-16.868+0-26.004-8.785-26.004-22.842zm63.429-.703v-31.803c-44.98+7.907-54.996+18.976-54.996+31.978+0+10.367+6.853+15.99+18.273+15.99+13.706.001+26.18-5.974+36.723-16.165z%5C%22%3E%3C%2Fpath%3E%5Cn+%5Ct%5Ct%5Ct%3Cpath+d%3D%5C%22M803.951+421.797v-92.772h15.286v13.881c8.083-7.907+19.68-15.813+34.79-15.813+17.746+0+27.41+10.191+27.41+27.762v66.943h-15.111v-63.604c0-12.299-5.271-18.098-17.043-18.098-11.069+0-20.382+5.798-30.046+14.935v66.768h-15.286zM903.572+375.411c0-31.978+20.382-48.319+43.224-48.319+22.666+0+43.048+16.341+43.048+48.319+0+31.802-20.382+48.319-43.048+48.319s-43.224-16.517-43.224-48.319zm70.986+0c0-19.328-9.313-35.316-27.762-35.316-17.746+0-27.937+14.408'
Stacktrace:
Error: ERR_FAILED (-2) loading 'file:///home/nig/dev/repositories/tutanota-3/build/index-desktop.html?noAutoLogin=true&platformId=linux&theme=%7B%22themeId%22%3A%22light%22%2C%22logo%22%3A%22%3Csvg+xmlns%3D%5C%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%5C%22+xml%3Aspace%3D%5C%22preserve%5C%22+preserveAspectRatio%3D%5C%22xMinYMid%5C%22+viewBox%3D%5C%220+0+1054.473+236.49%5C%22%3E%3Cg+transform%3D%5C%22translate%28-132.628+-222.799%29%5C%22+style%3D%5C%22fill%3A%234a4a4a%3Bfill-opacity%3A1%5C%22%3E%5Cn%5Ct%5Ct%5Ct%3Cpath+d%3D%5C%22M494.016+299.858H451.32v-8.082h94.178v8.082h-42.696v121.939h-8.785V299.858zM543.209+396.496v-66.943h8.434v65.889c0+14.057+6.15+20.908+20.206+20.908+12.827+0+23.544-6.676+34.965-17.57v-69.228h8.434v92.245h-8.434v-15.285c-10.191+9.662-22.139+17.219-35.668+17.219-19.152-.001-27.937-10.543-27.937-27.235zM652.319+401.24v-64.133h-16.165v-7.555h16.165v-33.56h8.434v33.56h24.071v7.555h-24.071v62.902c0+9.84+3.339+15.463+14.935+15.463+3.515+0+7.028-.352+9.664-1.23v7.906c-2.987.527-6.501.879-10.366.879-15.287+0-22.667-6.15-22.667-21.787zM703.27+400.537c0-17.57+14.935-30.749+63.43-38.831v-5.623c0-14.056-7.38-21.084-20.03-21.084-15.287+0-25.478+5.974-35.845+15.287l-4.919-5.271c11.421-10.542+23.192-17.395+40.938-17.395+19.152+0+28.289+10.894+28.289+27.937v43.399c0+11.246.703+18.097+2.636+22.841h-8.961c-1.23-3.865-2.108-8.434-2.108-13.705-11.245+9.664-23.545+15.287-37.426+15.287-16.868+0-26.004-8.785-26.004-22.842zm63.429-.703v-31.803c-44.98+7.907-54.996+18.976-54.996+31.978+0+10.367+6.853+15.99+18.273+15.99+13.706.001+26.18-5.974+36.723-16.165z%5C%22%3E%3C%2Fpath%3E%5Cn+%5Ct%5Ct%5Ct%3Cpath+d%3D%5C%22M803.951+421.797v-92.772h15.286v13.881c8.083-7.907+19.68-15.813+34.79-15.813+17.746+0+27.41+10.191+27.41+27.762v66.943h-15.111v-63.604c0-12.299-5.271-18.098-17.043-18.098-11.069+0-20.382+5.798-30.046+14.935v66.768h-15.286zM903.572+375.411c0-31.978+20.382-48.319+43.224-48.319+22.666+0+43.048+16.341+43.048+48.319+0+31.802-20.382+48.319-43.048+48.319s-43.224-16.517-43.224-48.319zm70.986+0c0-19.328-9.313-35.316-27.762-35.316-17.746+0-27.937+14.408'
at rejectAndCleanup (../node:electron/js2c/browser_init:161:7647)
at EventEmitter.stopLoadingListener (../node:electron/js2c/browser_init:161:8022)
at EventEmitter.emit (../node:events:390:28)
```

**Comment by charlag** [comment]
We don't get test failures and we handled multiple window scenario now I believe @ganthern

File: src/desktop/db/OfflineDbFacade.ts [hunk 1]
```diff
@@ -6,34 +6,56 @@ export interface OfflineDbFactory {
...
+			this.cache.delete(userId)
+		}
 	}
 
 	async deleteDatabaseForUser(userId: Id): Promise<void> {
-		// FIXME what if it's still open by another window?
-		await this.closeDatabaseForUser(userId)
+		const entry = this.cache.get(userId)
+		if (entry != null && entry.counter != 1) {
+			if (entry.counter != 1) {
```
Comment (ganthern) [inline src/desktop/db/OfflineDbFacade.ts:54]: pretty sure that's redundant

File: src/desktop/db/OfflineDbFacade.ts [hunk 2]
```diff
@@ -6,34 +6,56 @@ export interface OfflineDbFactory {
...
+				throw new Error(`Trying to delete database that is opened ${entry.counter} times`)
```
Comment (ganthern) [inline src/desktop/db/OfflineDbFacade.ts:55]: should this be a ProgrammingError?

File: src/desktop/db/OfflineDbFacade.ts [hunk 3]
```diff
@@ -6,34 +6,56 @@ export interface OfflineDbFactory {
...
 		}
 	}
 
 	async closeDatabaseForUser(userId: Id): Promise<void> {
-		const db = this.getDbForUserId(userId)
-		await db.close()
-		this.cache.delete(userId)
+		const entry = this.cache.get(userId)
+		if (entry == null) {
+			throw new Error(`Trying to close DB for user ${userId} but it was not open`)
```
Comment (ganthern) [inline src/desktop/db/OfflineDbFacade.ts]: should this be a ProgrammingError?

File: src/desktop/db/OfflineDbFacade.ts [hunk 4]
```diff
@@ -94,13 +116,13 @@ export class OfflineDbFacade {
...
 	}
 
-	private getDbForUserId(userId: Id,): OfflineDb {
-		const db = this.cache.get(userId)
+	private getDbForUserId(userId: Id): OfflineDb {
+		const entry = this.cache.get(userId)
 
-		if (!db) {
+		if (!entry) {
 			throw new Error(`Db for user ${userId} is not open. must call openDataBaseForUser first :)`)
```
Comment (ganthern) [inline src/desktop/db/OfflineDbFacade.ts:124]: should this be a ProgrammingError?

File: src/desktop/config/ConfigKeys.ts [hunk 5]
```diff
@@ -16,6 +16,7 @@ export enum DesktopConfigKey {
 	spellcheck = "spellcheck",
 	selectedTheme = "selectedTheme",
 	themes = "themes",
+	offlineStorage = "offlineStorage",
```
Comment (ganthern) [inline src/desktop/config/ConfigKeys.ts:19]: `offlineStorageEnabled` would be better

File: src/settings/DesktopSettingsViewer.ts [hunk 6]
```diff
@@ -212,31 +214,55 @@ export class DesktopSettingsViewer implements UpdatableSettingsViewer {
...
 		const defaultDownloadPathAttrs: TextFieldAttrs = {
 			label: "defaultDownloadPath_label",
 			value: this._defaultDownloadPath,
 			injectionsRight: () => m(ButtonN, changeDefaultDownloadPathAttrs),
 			disabled: true,
 		}
+<<<<<<< HEAD
+=======
+
+>>>>>>> 5d65a00d9 (Add config value for offline storage)
```
Comment (ganthern) [inline src/settings/DesktopSettingsViewer.ts:245]: w0t

File: src/settings/DesktopSettingsViewer.ts [hunk 7]
```diff
@@ -317,6 +346,8 @@ export class DesktopSettingsViewer implements UpdatableSettingsViewer {
 
 		this._spellCheckLang(spellcheckLabel)
 
+		this._offlienStorageValue(offlineStorage)
```
Comment (ganthern) [inline src/settings/DesktopSettingsViewer.ts:349]: typo: `offlien` -> `offline`

File: test/api/worker/facades/LoginFacadeTest.ts [hunk 8]
```diff
@@ -56,6 +56,7 @@ o.spec("LoginFacadeTest", function () {
 	let initializeCacheStorageMock: LateInitializedCacheStorage["initialize"]
 	let indexerMock: Indexer
 	let eventBusClientMock: EventBusClient
+	let usingOflfineStorage: boolean
```
Comment (ganthern) [inline test/api/worker/facades/LoginFacadeTest.ts:59]: typo: `oflfine` -> `offline`

File: src/api/common/Env.ts [hunk 9]
```diff
@@ -139,5 +139,6 @@ export function bootFinished() {
  * Whether or not we will be using an offline cache (doesn't take into account if credentials are stored)
  */
 export function isOfflineStorageAvailable(): boolean {
-	return false
+	// FIXME: make false again
```
[the remark this answers is not in the document]
Comment (charlag) [reply src/api/common/Env.ts:142]: We probably should already change it to `isDesktop()`

File: src/desktop/DesktopMain.ts [hunk 10]
```diff
@@ -140,19 +141,30 @@ async function createComponents(): Promise<Components> {
...
 
 	const offlineDbFacade = new OfflineDbFacade(offlineDbFactory)
 
-	const exposedInterfaceFactory = (windowId: number): ExposedNativeInterface => {
+	const exposedInterfaceFactory = (windowId: number, ipc: IPC): ExposedNativeInterface => {
 		return {
 			webauthn: new DesktopWebauthn(windowId, webDialogController),
-			offlineDbFacade: offlineDbFacade
+			offlineDbFacade,
+			interWindowEventSender: new DesktopInterWindowEventSender(ipc, wm),
```
[the remark this answers is not in the document]
Comment (charlag) [reply src/desktop/DesktopMain.ts]: It doesn't define what is sent over the bus, just the bus itself. Communication between windows is something specific to desktop, is it not? Where could we put it?

[the remark this answers is not in the document]
Comment (charlag) [reply src/desktop/DesktopMain.ts]: WindowManager doesn't need to know about bus either

[the remark this answers is not in the document]
Comment (ganthern) [reply src/desktop/DesktopMain.ts]: not particularly, I just feel that as soon as you use the ApplicationWindow class they should be able to use the bus without further configuration.
the wm could connect the windows to the bus when it creates them. maybe IPC is a better candidate.

we currently have a kind of messy circular relationship between ipc and wm.

[the remark this answers is not in the document]
Comment (charlag) [reply src/desktop/DesktopMain.ts]: Well we don't implement eventBus in DesktopMain, we just connect the pieces like it was doing before. We need IPC rewrite regardless, I don't know if I want to invest time into doing it now.

[the remark this answers is not in the document]
Comment (ganthern) [reply src/desktop/DesktopMain.ts]: then let's just say that the bus should be part of the appwindow/ipc module if we ever get to untangling it

**Comment by charlag** [comment]
Oops test failed

**Comment by ganthern** [comment]
my local test failure cleared up after deleting native-cache and rebuilding sqlite.

**Review by ganthern — APPROVED** [review]

**Pull request merged**